examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+                 scratch_data.fe_values.JxW(q_point)) +
+                ((advection_direction *
+                  scratch_data.fe_values.shape_grad(j, q_point)) *
+                 scratch_data.fe_values.shape_value(i, q_point)) *
+                  scratch_data.fe_values.JxW(q_point);
+
+              if (settings.with_sd)
+                copy_data.cell_matrix(i, j) +=
+                  delta *
+                    (advection_direction *
+                     scratch_data.fe_values.shape_grad(j, q_point)) *
+                    (advection_direction *
+                     scratch_data.fe_values.shape_grad(i, q_point)) *
+                    scratch_data.fe_values.JxW(q_point) -
+                  delta * epsilon *
+                    trace(scratch_data.fe_values.shape_hessian(j, q_point)) *
+                    (advection_direction *
+                     scratch_data.fe_values.shape_grad(i, q_point)) *
+                    scratch_data.fe_values.JxW(q_point);
+            }
+          if (!cell->is_level_cell())
+            {
+              copy_data.cell_rhs(i) +=
+                scratch_data.fe_values.shape_value(i, q_point) *
+                rhs_values[q_point] * scratch_data.fe_values.JxW(q_point);
+              if (settings.with_sd)
+                copy_data.cell_rhs(i) +=
+                  delta * rhs_values[q_point] * advection_direction *
+                  scratch_data.fe_values.shape_grad(i, q_point) *
+                  scratch_data.fe_values.JxW(q_point);
+            }
+        }
+  }
+
+
+  template <int dim>
+  void AdvectionProblem<dim>::assemble_system_and_multigrid()
+  {
+    auto cell_worker_active =
+      [&](const decltype(dof_handler.begin_active()) &cell,
+          ScratchData<dim> &                          scratch_data,
+          CopyData &                                  copy_data) {
+        this->assemble_cell(cell, scratch_data, copy_data);
+      };
+
+
+
+    auto copier_active = [&](const CopyData &copy_data) {
+      constraints.distribute_local_to_global(copy_data.cell_matrix,
+                                             copy_data.cell_rhs,
+                                             copy_data.local_dof_indices,
+                                             system_matrix,
+                                             system_rhs);
+    };
+
+
+    MeshWorker::mesh_loop(dof_handler.begin_active(),
+                          dof_handler.end(),
+                          cell_worker_active,
+                          copier_active,
+                          ScratchData<dim>(fe, quad_degree),
+                          CopyData(),
+                          MeshWorker::assemble_own_cells);
+
+
+    // Assemble GMG
+    std::vector<AffineConstraints<double>> boundary_constraints(
+      triangulation.n_global_levels());
+    for (unsigned int level = 0; level < triangulation.n_global_levels();
+         ++level)
+      {
+        IndexSet dofset;
```

Comment:
let's use a better name -- maybe `locally_owned_level_dof_indices`? It's a long name, but at least accurately describes what it is.